activerecord/lib/active_record/connection_adapters/mysql2_adapter.rb

```diff
@@ -120,12 +115,12 @@ def quote_string(string)
       #++
 
       def active?
-        @raw_connection.ping
+        !!@raw_connection&.ping
       end
 
       def reconnect!(restore_transactions: false)
         @lock.synchronize do
-          @raw_connection.close
+          @raw_connection&.close
```

Comment:
I don't like that now we have to keep checking for `nil` everywhere. I'm concerned we'll forget to do this in the future for a new method.

Reply:
I think it's fairly fine... we have plenty of potentially-nil ivars in various places. The main mitigating factor here is that we should rarely add new methods that use the variable directly: outside of connection management (dis/re-/connect), nearly everything else should use `with_raw_connection`, `valid_raw_connection`, or `any_raw_connection` (in that preference order).

The other side is just that there isn't much of an alternative: while abstract adapter defines that `@raw_connection` exists, is nil when not connected, and truthy when connected... the rest of the API is adapter-specific, so it's impossible for us to use a null-object pattern, say.